cpp/unreal_plugins/UrdfBot/Source/UrdfParser.cpp

```diff
@@ -0,0 +1,415 @@
+        joint_desc.type_ = UrdfJointType::Planar;
+    } else {
+        ASSERT(false);
+    }
+
+    bool has_origin_node = false;
+    bool has_parent_node = false;
+    bool has_child_node = false;
+    bool has_axis_node = false;
+    bool has_calibration_node = false;
+    bool has_dynamics_node = false;
+    bool has_limit_node = false;
+    bool has_mimic_node = false;
+    bool has_safety_controller_node = false;
+
+    for (auto& node : joint_node->GetChildrenNodes()) {
+        const FString& tag = node->GetTag();
+        if (tag.Equals(TEXT("origin"))) {
+            ASSERT(!has_origin_node);
+            has_origin_node = true;
+            joint_desc.origin_.SetLocation(parseVector(node->GetAttribute("xyz")));
+            joint_desc.origin_.SetRotation(parseRotation(node->GetAttribute("rpy")));
+        } else if (tag.Equals(TEXT("parent"))) {
+            ASSERT(!has_parent_node);
+            has_parent_node = true;
+            joint_desc.parent_ = TCHAR_TO_UTF8(*node->GetAttribute(TEXT("link")));
+        } else if (tag.Equals(TEXT("child"))) {
+            ASSERT(!has_child_node);
+            has_child_node = true;
+            joint_desc.child_ = TCHAR_TO_UTF8(*node->GetAttribute(TEXT("link")));
+        } else if (tag.Equals(TEXT("axis"))) {
+            ASSERT(!has_axis_node);
+            has_axis_node = true;
+            joint_desc.axis_ = parseVector(node->GetAttribute(TEXT("xyz")));
+        } else if (tag.Equals(TEXT("calibration"))) {
+            ASSERT(!has_calibration_node);
+            has_calibration_node = true;
+            for (auto& attr_node : node->GetAttributes()) {
+                if (attr_node.GetTag() == "rising") {
+                    ASSERT(joint_desc.calibration_type_ == CalibrationType::Invalid);
+                    joint_desc.calibration_type_ = CalibrationType::Rising;
+                    joint_desc.rising_ = FCString::Atof(*attr_node.GetValue());
+                } else if (attr_node.GetTag() == "falling") {
+                    ASSERT(joint_desc.calibration_type_ == CalibrationType::Invalid);
+                    joint_desc.calibration_type_ = CalibrationType::Falling;
+                    joint_desc.falling_ = FCString::Atof(*attr_node.GetValue());
+                }
+            }
+        } else if (tag.Equals(TEXT("dynamics"))) {
+            ASSERT(!has_dynamics_node);
+            has_dynamics_node = true;
+            joint_desc.damping_ = FCString::Atof(*node->GetAttribute(TEXT("damping")));
+            joint_desc.friction_ = FCString::Atof(*node->GetAttribute(TEXT("friction")));
+        } else if (tag.Equals(TEXT("limit"))) {
+            ASSERT(!has_limit_node);
+            has_limit_node = true;
+            joint_desc.lower_ = FCString::Atof(*node->GetAttribute(TEXT("lower")));
+            joint_desc.upper_ = FCString::Atof(*node->GetAttribute(TEXT("upper")));
+            joint_desc.velocity_ = FCString::Atof(*node->GetAttribute(TEXT("velocity")));
```

Comment:
- `lower` and `upper` are listed as optional in the spec. Will this code work as expected if they're omitted?
- swap the order of `velocity` and `effort` to match the spec.

Reply:
fixed